### PR TITLE
article: add part 3 of DReddit tutorial series

### PR DESCRIPTION
**Pull request by PascalPrecht**

### CODE REVIEW

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+  });
+
+  this.setState({ posts: list });
+}</code></pre>
+
+That's it! Our application now renders a list of all created posts. Unfortunately, posts are not being re-fetched  automatically when adding new posts. For the time being, we'll have to reload the browser every time after adding a post. However, this we'll address now.
+
+### Reloading posts
+
+There is certainly different ways to make the list of posts update automatically, so take the following approach with a grain of salt. What we need is a way to have the `createPost` component tell the `List` component to reload its posts. However, there's not communication layer in place when building a simple React app like this, so the most straight forward way to make this possible, is to move the logic of loading the posts in the parent component of `CreatePost` and `List` (in our case `App`), and have it pass that logic down to places where its needed. This also means we'll be fetching the list inside `App` and pass down the pure data to `List`.
```
Comment (jrainville) [inline source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]:
```suggestion
There is certainly different ways to make the list of posts update automatically, so take the following approach with a grain of salt. What we need is a way to have the `createPost` component tell the `List` component to reload its posts. However, there's no communication layer in place when building a simple React app like this, so the most straight forward way to make this possible, is to move the logic of loading the posts in the parent component of `CreatePost` and `List` (in our case `App`), and have it pass that logic down to places where its needed. This also means we'll be fetching the list inside `App` and pass down the pure data to `List`.
```

Comment (jrainville) [reply source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]: Also, I just thought, there is another way to do it. You can always use contract `events` in the List and listen to new posts.

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+    };
+  }
+
+  async componentDidMount() {
+    await this.loadPosts();
+  }
+  ...
+}</code></pre>
+
+Last but not least, all we have to do is passing the `posts` down to `List` and `loadPosts()` to `CreatePost` as a callback handler if you will:
```
Comment (jrainville) [inline source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]: You use "Last but not least" a lot. Maybe changing it from time to time.

Comment (jrainville) [reply source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]:
```suggestion
Last but not least, all we have to do is pass the `posts` down to `List` and `loadPosts()` to `CreatePost` as a callback handler if you will:
```

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+    &lt;React.Fragment&gt;
+      &lt;h1&gt;DReddit&lt;/h1&gt;
+      &lt;CreatePost afterPostHandler={this.loadPosts.bind(this)}/&gt;
+      &lt;List posts={this.state.posts}/&gt;
+    &lt;/React.Fragment&gt;
+  )
+}
+</code></pre>
+
+Once that is done, we can consume `posts` and `afterPostHandler()` from `this.props` respectively. In `List` component's `render()` function we'll do (notice we don't rely on `this.state` anymore):
```
Comment (jrainville) [inline source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]:
```suggestion
Once that is done, we can consume `posts` and `afterPostHandler()` from `this.props` respectively. In `List`'s `render()` function we'll do (notice we don't rely on `this.state` anymore):
```

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+          downvotes={post.downvotes}
+          owner={post.owner}
+          /&gt;)
+      })}
+      &lt;/React.Fragment&gt;
+    )
+  }
+}</code></pre>
+
+Rendering the number of `upvotes` and `downvotes` is then really just a matter of interpolating them in `Post` component's `render()` function. We'll just going to add them next to the buttons, but feel free to put them somewhere else:
```
Comment (jrainville) [inline source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]:
```suggestion
Rendering the number of `upvotes` and `downvotes` is then really just a matter of interpolating them in `Post`'s `render()` function. We're just going to add them next to the buttons, but feel free to put them somewhere else:
```

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+          &lt;textarea name="content"&gt;&lt;/textarea&gt;
+        &lt;/div&gt;
+        &lt;button&gt;Post&lt;/button&gt;
+      &lt;/form&gt;
+    )
+  }
+}
+</code></pre>
+
+To actually render this component on screen, we need to make it part of our `App` component. Or, to be more specific, have the `App` component render our `CreatePost` component. For no we can simply extend `App`'s render function like this;
```
[the remark this answers is not in the document]
Comment (PascalPrecht) [reply source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]: 👍

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+  }));
+
+  this.setState({
+    topic: '',
+    content: '',
+    loading: false
+  });
+}</code></pre>
+
+We use `JSON.stringify()` on an object that holds the `topic` and `content` of the post to be created. This is also the first time we put `loading` into action. Setting it to `true` before, and `false` after we've performed our operations lets us render a useful message to the user as she's waiting for updates.
```
[the remark this answers is not in the document]
Comment (PascalPrecht) [reply source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]: Ha, and here I thought I'd do something good for being inclusive :D

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+Next up we'll import a `DReddit` Smart Contract instance from EmbarkJS and use it to get a gas estimation from `web3`. We can then use the estimation and one of our accounts to actually send the transaction:
+
+
+<pre class="javascript"><code>import DReddit from 'Embark/contracts/DReddit';
+...
+
+async createPost(event) {
+  ...
+  const accounts = await web3.eth.getAccounts();
+  const createPost = await DReddit.methods.createPost(web3.utils.toHex(ipfsHash));
```
[the remark this answers is not in the document]
Comment (PascalPrecht) [reply source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]: I think I've addressed this in the paragraph before that code snippet

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+
+export class Post extends Component {
+
+  constructor(props) {
+    super(props);
+  }
+
+  async componentDidMount() {
+    const ipfsHash = web3.utils.toAscii(this.props.description);
+    const data = await EmbarkJS.Storage.get(ipfsHash);
```
[the remark this answers is not in the document]
Comment (PascalPrecht) [reply source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]: 👍 good catcH!

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+Let's fill our new `List` component with life! As mentioned earlier, we'll use our Smart Contract's `numPosts()` method to get hold of the total number of posts available. We then use that number to iterate over all indices and request every post individually. Since this is logic we want to execute once `List` component is ready, we'll use its `componentDidMount()` method for that:
+
+<pre class="javascript"><code>export class List extends Component {
+  ...
+  async componentDidMount() {
+    const totalPosts = await DReddit.methods.numPosts().call();
+
+    let list = [];
+
+    if (totalPosts > 0) {
```
[the remark this answers is not in the document]
Comment (PascalPrecht) [reply source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]: Yes!

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md (hunk 1 above)
Comment (PascalPrecht) [reply source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]: Yea good point!

File: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jrainville) [reply source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md]: Haha, yeah, I didn't want to sound like a cis-male. Just that it's the only place we use `she` so it felt random. I change dthe sentence so use neither so we don't frustrate anyone haha

**Pull request merged**